docs/developers/testing.md

```diff
@@ -0,0 +1,314 @@
+
+The C++ source file should be added to the `src` directory in `plugins/microservices`, and the test microservice name added to the list of `IRODS_MICROSERVICE_PLUGINS` in the CMakeLists file.
+
+#### Add a Python test for running the test microservice
+
+Here is the form for implementing a Python test which runs one of the server-side microservice tests:
+```python
+# Consider skipping the test for non-iRODS Rule Language REPs, or implementing for those other REPs.
+def test_my_server_side_only_feature(self):
+    rep_name = 'irods_rule_engine_plugin-irods_rule_language-instance'
+    self.admin.assert_icommand(
+        ['irule', '-r', rep_name, 'msi_test_my_server_side_only_feature', 'null', 'ruleExecOut'])
+```
+The test only asserts that the `irule` invocation does not emit any errors. This is asserting that all of the assertions in the test microservice are passing.
+
+Add the Python test to an existing Python test suite found in the core test list, or create a new Python test file and add the name to the core test list to ensure that it is run along with all the other tests. See [Core test suite](#core-test-suite) for more information about adding and running Python tests.
+
+## iRODS Testing Environment
+
+The iRODS Testing Environment is a Docker Compose-based project written in Python that is used by the Consortium to test the iRODS server and its plugins. The Testing Environment is driven by a series of scripts to run the tests on a number of different platforms with various database backends and other configuration options. This is the primary tool used for day-to-day testing for iRODS developers as well as for verification in releasing the iRODS server. In addition to running tests, the Compose projects can be useful for manually testing things and quickly deploying specific environments for reproducing issues or developing a proof of concept.
+
+This documentation will briefly describe a few useful ways to employ the iRODS Testing Environment. The README in the project's GitHub repository provides extensive details on the purpose and usage for the various scripts: [https://github.com/irods/irods_testing_environment/blob/main/README.md](https://github.com/irods/irods_testing_environment/blob/main/README.md) Each script available in the project also has help text available with the `--help` option which describes the available options and their purposes.
+
+### Getting started
+
+In order to configure your environment to use the iRODS Testing Environment, using a virtual environment is highly recommended. Here's how to set up a virtual environment and install the required Python packages:
+```
+virtualenv -p python3 ~/irods_testing_environment
+source ~/irods_testing_environment/bin/activate
+pip install docker-compose GitPython
+pip freeze
```

Comment:
why pip freeze?  is this just to show that the things are there...

Reply:
I've removed the line, so I'm going to resolve this.